sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignature.java

```diff
@@ -114,11 +114,116 @@ static Builder builder() {
   }
 
   /** A type of optional parameter of the {@link DoFn.ProcessElement} method. */
-  public enum Parameter {
-    BOUNDED_WINDOW,
-    INPUT_PROVIDER,
-    OUTPUT_RECEIVER,
-    RESTRICTION_TRACKER
+  public static abstract class Parameter {
```

Comment:
This switches `Parameter` from an enum to something like an "Object Algebra", which is a visitor-based church-encoding of a disjoint union. It actually incurs a lot of boilerplate. The other popular disjoint union encoding is to just do pattern matching via explicit `instanceof` instead of the visitor.

Note that a standard object algebra does not have the `visit` method accept subclasses, but would have each of them parameterized by payloads, so it would be:

```java
public interface Visitor<ResultT> {
  ResultT boundedWindow();
  ResultT restrictionTracker();
  ResultT state(String id, StateDeclaration decl);
}
```

This makes static factory methods isomorphic to the visitor, which is kind of nice.